test/functional/unified-spec-runner/entities.ts

```diff
@@ -38,52 +62,105 @@ function getClient(address) {
   return new MongoClient(`mongodb://${address}`, serverApi ? { serverApi } : {});
 }
 
+type PushFunction = (e: CommandEvent | CmapEvent) => void;
+
 export class UnifiedMongoClient extends MongoClient {
-  events: CommandEvent[];
+  commandEvents: CommandEvent[];
+  cmapEvents: CmapEvent[];
   failPoints: Document[];
   ignoredEvents: string[];
-  observedEvents: ('commandStarted' | 'commandSucceeded' | 'commandFailed')[];
+  observedCommandEvents: ('commandStarted' | 'commandSucceeded' | 'commandFailed')[];
+  observedCmapEvents: (
+    | 'connectionPoolCreated'
+    | 'connectionPoolClosed'
+    | 'connectionPoolCleared'
+    | 'connectionCreated'
+    | 'connectionReady'
+    | 'connectionClosed'
+    | 'connectionCheckOutStarted'
+    | 'connectionCheckOutFailed'
+    | 'connectionCheckedOut'
+    | 'connectionCheckedIn'
+  )[];
 
-  static EVENT_NAME_LOOKUP = {
+  static COMMAND_EVENT_NAME_LOOKUP = {
     commandStartedEvent: 'commandStarted',
     commandSucceededEvent: 'commandSucceeded',
     commandFailedEvent: 'commandFailed'
   } as const;
 
+  static CMAP_EVENT_NAME_LOOKUP = {
+    poolCreatedEvent: 'connectionPoolCreated',
+    poolClosedEvent: 'connectionPoolClosed',
+    poolClearedEvent: 'connectionPoolCleared',
+    connectionCreatedEvent: 'connectionCreated',
+    connectionReadyEvent: 'connectionReady',
+    connectionClosedEvent: 'connectionClosed',
+    connectionCheckOutStartedEvent: 'connectionCheckOutStarted',
+    connectionCheckOutFailedEvent: 'connectionCheckOutFailed',
+    connectionCheckedOutEvent: 'connectionCheckedOut',
+    connectionCheckedInEvent: 'connectionCheckedIn'
+  } as const;
+
   constructor(url: string, description: ClientEntity) {
     super(url, {
       monitorCommands: true,
       ...description.uriOptions,
       serverApi: description.serverApi ? description.serverApi : serverApiConfig()
     });
-    this.events = [];
+    this.commandEvents = [];
+    this.cmapEvents = [];
     this.failPoints = [];
     this.ignoredEvents = [
       ...(description.ignoreCommandMonitoringEvents ?? []),
       'configureFailPoint'
     ];
-    // apm
-    this.observedEvents = (description.observeEvents ?? []).map(
-      e => UnifiedMongoClient.EVENT_NAME_LOOKUP[e]
+    this.observedCommandEvents = (description.observeEvents ?? []).map(
```

Comment:
should these two observed event arrays be filtered so that only the defined ones get listeners since they both derive from `description.observeEvents`?

Reply:
These are now filtered so no `undefined` elements are in the arrays.